### DISCUSSION TITLE
Determine at runtime if all the test cases were run

### DISCUSSION THREAD

**Discussion by CraigSiemens** [Quick/Quick#1296] [Q&A]
Is there any way with either Quick or XCTest to determine if any test is skipped for any reason. The possibilities I can think of are:
- running a single test/file/target from Xcode
- focused/disabled examples in Quick (`fit`, `xit`, etc.)
- (maybe?) if the tests are run in parallel

## The issue behind the question

I our project we use https://github.com/pointfreeco/swift-snapshot-testing for testing some of our views. When we rename or remove a test, sometimes we forget to remove the corresponding snapshot the test.

The idea is that it might be possible to register the snapshots that were used by the tests, then at the end warn or remove any snapshots that were unused. For this to be accurate, it'd need to know whether a snapshot was unused because the test was skipped vs not existing.

**Comment by younata** [comment]
There's a couple ways to do this:

1. You could parse the output, whether that's the raw console output, or the xml file that something like `fastlane scan` will optionally output. Then you can look for any test marked as skipped. If I were to do this, then I would go with this approach.
2. If you absolutely need to do this at runtime, you can utilize the [XCTestObservation](https://developer.apple.com/documentation/xctest/xctestobservation) infrastructure, and register an observer that implements [`testSuitDidFinish`](https://developer.apple.com/documentation/xctest/xctestobservation/1500958-testsuitedidfinish). This observer can then iterate over each test's testRun property, and check if `hasBeenSkipped` is true. Some sample code that I haven't tried to compile:

```swift
final class SkippedTestObserver: NSObject, XCTestObservation {
    func testSuiteDidFinish(_ testSuite: XCTestSuite) {
        let skippedTests: [XCTest] = testSuite.tests.filter { $0.testRun?.hasBeenSkipped == true }
        // TODO: log the skipped tests or something?
    }
}
// TODO: register an instance of SkippedTestObserver with XCTestObservationCenter somewhere.
```

Again, you probably don't care to output skipped test after each run - this code won't take long to run, but it will clutter your console output - it's probably better to write a separate script that parses the console (or xml) output and detects skipped tests. Then you can invoke that script only when you care to look for that.
